Add postgres-specific id generator

4word opened the pull request:
Requiring an external-to-the-db id generator is not an option for some use cases. This new field and related migration utilities allow for shard-safe (up to 8191 shards) inserts with guaranteed unique IDs (up to 1024 inserts per millisecond on an individual shard). The ID will return like a standard Django auto field ID. In the future, a given ID could be deconstructed to get the time it was inserted and the shard id (useful for the router to potentially know which shard to grab something from without needing help from the using() statement).

4word commented:
Note that much of this code was ported over from a Django 1.4 and python 2.7 deployment and I am not exactly expecting unit tests to pass the first time (although I updated as much as I could think of and find).

4word commented:
This is ready for more scrutinizing review. Docs need to be updated but functionality is available and tested for both database and django level usage.

JBKahn commented on django_sharding_library/fields.py:
```diff
@@ -156,3 +159,30 @@ class ShardForeignKeyStorageField(ShardForeignKeyStorageFieldMixin, ForeignKey):
+        if not db_alias:
+            raise EnvironmentError("A pre-migration receiver did not receive a database alias. "
+                                   "Perhaps your app is not registered correctly?")
+        if settings.DATABASES[db_alias]['ENGINE'] == Backends.POSTGRES:
```
since this only supports postgres, somewhere (maybe here?) we should raise an exception if you try to use it and you're not using postgres.

4word replied:
Hmm. The problem is that the library allows arbitrary backends within the same shard group. Should I just loop through the entire shard group and make sure all the backends are postgres?

JBKahn replied:
Yup, you can do it either for all sharded databases or make it a shard
group level setting?

On Tue, May 3, 2016, 8:59 PM Titus Peterson <notifications@github.com>
wrote:

> In django_sharding_library/fields.py
> <https://github.com/JBKahn/django-sharding/pull/26#discussion_r61980198>:
>
> > +        if not hasattr(settings, 'SHARD_EPOCH'):
> > +            raise ValueError("PostgresShardGeneratedIDField requires a SHARD_EPOCH to be defined in your settings file.")
> > +
> > +        if connection.vendor == PostgresDatabaseWrapper.vendor:
> > +            return "bigint DEFAULT next_sharded_id()"
> > +        else:
> > +            return super(PostgresShardGeneratedIDField, self).db_type(connection)
> > +
> > +    @staticmethod
> > +    def migration_receiver(*args, **kwargs):
> > +        sequence_name = "global_id_sequence"
> > +        db_alias = kwargs.get('using')
> > +        if not db_alias:
> > +            raise EnvironmentError("A pre-migration receiver did not receive a database alias. "
> > +                                   "Perhaps your app is not registered correctly?")
> > +        if settings.DATABASES[db_alias]['ENGINE'] == Backends.POSTGRES:
>
> Hmm. The problem is that the library allows arbitrary backends within the
> same shard group. Should I just loop through the entire shard group and
> make sure all the backends are postgres?
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/JBKahn/django-sharding/pull/26/files/e534b5ffa93bd79fab7bcb8f90fb315a344f3661#r61980198>
>

4word replied:
Updated to raise an exception

JBKahn commented on django_sharding_library/utils.py:
```diff
@@ -0,0 +1,32 @@
+
+
+def register_migration_signal_for_model_receiver(model, function, dispatch_uid=None):
+    signals.pre_migrate.connect(function, sender=model, dispatch_uid=dispatch_uid)
```
why pre_migrate? Is there a check that it was successful or a chance there will be a silent failure such that one node doesn't have it?

4word replied:
It will get a DatabaseError if it fails. Its in pre-migrate because its the only common area to add a migration step without editing the migration file itself after its produced, which I don't think is a good feature for a library since it adds complexity to using the library that can be taken care of automatically in the pre migration.

Also, it has to be in the pre migrate because if one of the postgres fields gets created and the stored procedure doesnt exist yet, it will raise a programmingerror. Doing it in a post migration or anywhere else doesn't work.

JBKahn replied:
Hmm definitely don't think we need to force users to modify sharded tables.
If we can, I have to look into how database migration dependencies work but
could use a migration dependency on a migration shipped with the app?

On Tue, May 3, 2016, 9:02 PM Titus Peterson <notifications@github.com>
wrote:

> In django_sharding_library/utils.py
> <https://github.com/JBKahn/django-sharding/pull/26#discussion_r61980374>:
>
> > +            cursor.execute("SELECT setval('%s', 1, false)" % (sequence_name,))
> > +    else:
> > +        transaction.savepoint_commit(sid, using=db_alias)
> > +    cursor.close()
> > +
> > +
> > +def create_postgres_shard_id_function(sequence_name, db_alias, shard_id):
> > +    cursor = connections[db_alias].cursor()
> > +    cursor.execute(postgres_shard_id_function_sql % {'shard_epoch': settings.SHARD_EPOCH,
> > +                                                     'shard_id': shard_id,
> > +                                                     'sequence_name': sequence_name})
> > +    cursor.close()
> > +
> > +
> > +def register_migration_signal_for_model_receiver(model, function, dispatch_uid=None):
> > +    signals.pre_migrate.connect(function, sender=model, dispatch_uid=dispatch_uid)
>
> It will get a DatabaseError if it fails. Its in pre-migrate because its
> the only common area to add a migration step without editing the migration
> file itself after its produced, which I don't think is a good feature for a
> library since it adds complexity to using the library that can be taken
> care of automatically in the pre migration.
>
> Also, it has to be in the pre migrate because if one of the postgres
> fields gets created and the stored procedure doesnt exist yet, it will
> raise a programmingerror. Doing it in a post migration or anywhere else
> doesn't work.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/JBKahn/django-sharding/pull/26/files/e534b5ffa93bd79fab7bcb8f90fb315a344f3661#r61980374>
>

4word replied:
As far as I can find in the documentation, there is no way to force a dependency because of a field when a user run makemigrations. I also asked in the django irc and the response from the admin there at the time was that he didn't think it was possible.

This is exactly why the pre-migration signal was added in though, its for pre-migration processing. I am pretty sure this is the "django" way of doing it short of writing a migration, putting it in the app, and figuring out how to force it as a dependency on another app based on arbitrary, non-built-in logic (for example, foreign key fields have special logic already built into the migration builder). I am not sure I can think of a better use case of the pre migration signal than this

4word replied:
Let me know if you find something to handle a dependency, I actually already have a migration written to do this that I could include if we figure it out (in my current project, I was adding the migration manually to the automated migration files before I wrote this)

JBKahn replied:
Like I said, I have no ideas, make a special note in your docs about this. Having to run this. Ideally I'd also like to add a management command to run this, as well as one to check that ti was run. Since the django way feels a bit hacky to me.

4word replied:
I updated the docs to indicate this is happening. I can add a migrate command in the future: this is pretty failure-safe. If for any reason the required function or sequence do not exist, postgres itself will error out and send the error message to Django. But I agree, a manage command is in order, at least for checking that the function exists and is working properly. Will do that in a future pull request.

JBKahn replied:
add the management command in this one, at least a rudimentary one.
i.e.
```python
ipdb> cursor.execute("SELECT count(*) FROM pg_class c WHERE c.relkind = '%s' and c.relname = '%s';" % ('S', sequence_name))
ipdb> cursor.fetchone()
(1L,)
```
and just take a sequence name (or whatever requirements to generate the ones you use) and the database to check. As well as one to create the sequence and to reset it, using the code you already wrote.

JBKahn replied:
well I supose it can be in a different PR but I don't want to cut a release till it's in.

4word commented:
Updated the decorator to raise an error when the postgres field is used on a non-postgres database.

4word commented:
Fixed a bug where when there are multiple models in the same app, and some of them do not use postgres, the migration was trying to run the postgres-specific functions on the non-postgres models. Its fixed to check for postgres correctly now.

JBKahn commented:
I'm going to spend one more day stewing on it, the dependancy code didn't seem to offer a lot of ways to reasonably hook into it.

JBKahn commented on django_sharding_library/decorators.py:
```diff
@@ -24,12 +30,30 @@ def configure(cls):
+        postgres_shard_id_fields = list(filter(lambda field: issubclass(type(field), PostgresShardGeneratedIDField), cls._meta.fields))
+        if postgres_shard_id_fields:
+            # Make sure we are only using postgres
+            if database:
```
I'd just refactor that check into something like:
````python
if postgres_shard_id_fields:
    database_dicts = [settings.DATABASES[database]] if database else [db_settings for db, db_settings in viewitems(settings.DATABASES) if db_settings["SHARD_GROUP"] == shard_group]
    if any([database_dict['ENGINE'] not in Backends.POSTGRES for database_dict in database_dicts]):
        raise ShardedModelInitializationException('You cannot use a PostgresShardGeneratedIDField on a non-Postgres database.')

4word replied:
Good call, that looks cleaner

JBKahn commented on django_sharding_library/decorators.py:
```diff
@@ -24,12 +30,30 @@ def configure(cls):
+                        raise ShardedModelInitializationException('You cannot use a PostgresShardGeneratedIDField on a '
+                                                                  'non-Postgres database.')
+
+            register_migration_signal_for_model_receiver(apps.get_app_config(cls._meta.app_label),
```
I spent more time reading that code for dependencies and it isn't worth trying to hack into it, given they don't expose any good hooks for us to use.

4word replied:
Yeah its unfortunate that django has opted to keep migrations behind a closed curtain it seems like. Will look at updating the docs today.

JBKahn commented on django_sharding_library/decorators.py:
```diff
@@ -24,12 +30,30 @@ def configure(cls):
         if shard_group:
             sharded_fields = list(filter(lambda field: issubclass(type(field), ShardedIDFieldMixin), cls._meta.fields))
-            if not sharded_fields:
+            if not sharded_fields and not postgres_shard_id_fields:
```
update those two exception messages to include that using the `PostgresShardGeneratedIDField` is also valid.

4word replied:
Will do that today

JBKahn commented on django_sharding_library/fields.py:
```diff
@@ -156,3 +159,30 @@ class ShardForeignKeyStorageField(ShardForeignKeyStorageFieldMixin, ForeignKey):
+
+    @staticmethod
+    def migration_receiver(*args, **kwargs):
+        sequence_name = "global_id_sequence"
```
Ohh it's one sequence per DB,, as opposed to per model?

4word replied:
Correct, its per DB. There is no reason to have a per model, no DB (that I am aware of) can insert more than 1024 rows per millisecond (which is what this id generator is designed to be able to do)!

JBKahn replied:
when we were deciding on that strategy, we just weren't sure about id growth and sharing them, although now that I consider the actual size of a bigint, I have no idea why the decision was made not to do that.

JBKahn commented on django_sharding_library/settings_helpers.py:
```diff
@@ -89,4 +90,13 @@ def database_configs(databases_dict):
+
+            # We assume the numeric shard ID is constant based on the entries in the configuration helper (we assume
+            # they wont change order, and that new shards will be appended and not inserted randomly)
+            # todo: update docs to make sure this assumption is understood
```
very important if we make that assumption.

4word replied:
I will make sure to have that in the doc

4word replied:
Doc updated with this assumption

JBKahn replied:
Awesome, a bit out of commission for a few days but later this week I'll
get the Travis pull request merged as well as take another look at your code

On Sun, May 15, 2016, 7:50 PM Titus Peterson <notifications@github.com>
wrote:

> In django_sharding_library/settings_helpers.py
> <https://github.com/JBKahn/django-sharding/pull/26#discussion_r63300509>:
>
> > @@ -89,4 +90,13 @@ def database_configs(databases_dict):
> >                  )
> >                  if db_config:
> >                      configuration[replica['name']] = db_config
> > +
> > +            # We assume the numeric shard ID is constant based on the entries in the configuration helper (we assume
> > +            # they wont change order, and that new shards will be appended and not inserted randomly)
> > +            # todo: update docs to make sure this assumption is understood
>
> Doc updated with this assumption
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/JBKahn/django-sharding/pull/26/files/4e531964efbf77ff8c4fb4d5a57829a49a1b521c#r63300509>
>

4word commented:
Will be trying to wrap this up later today, I tried to get to it earlier but it was really busy this week.

JBKahn commented:
The code looks good, two questions:
1) Is there an upper limit to the number of shards given they are assigned ids, although I assume it's very large in either event?
2) If the shard epoch can never be changed anyway, is there a reason we don't set a default that we let you override? I assume that unless you're moving from basically the same system to this one, you'd never care about the value?

4word commented:
There is an upper limit, it's stored in 10 bits which limits it to 8192 (minus 1, since the code assumes you start your shard numbering at 1 instead of 0) total shards. I will put that in the documentation.

I can put a default in there, the main thing is that anyone writing code in the future needs to never change the default value because it is backwards incompatible, anyone who starts using the library with the default will never be able to change it without migrating all their data.

Because of the gravity of the potential amount of work involved with changing the epoch, I would rather just raise an exception and make the person make a decision about it. They can use one of the other ID generators in the library if they don't want to think about it.

JBKahn commented:
that's reasonable. I'll merge this one in, but I'll hold off on the release till the management command is in since I'd rather group those into a release.

JBKahn commented:
Well, after you modify the docs :)

4word commented:
That sounds good. I'll update the docs later today.
On Sun, May 29, 2016 at 1:06 PM Joseph Kahn <notifications@github.com>
wrote:

> that's reasonable. I'll merge this one in, but I'll hold off on the
> release till the management command is in since I'd rather group those into
> a release.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/JBKahn/django-sharding/pull/26#issuecomment-222382703>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AB4RuiYvSfNCYiwq61xfYV4Tjzb6dsRsks5qGf_AgaJpZM4IV4It>
> .
>

4word commented:
I rebased my commits to squash down the sheer volume of them that there were. I updated the doc to mention the shard limit.

This is ready to merge in, I am planning on doing the management commands in a new PR. I would rather wait for you to merge this in until after you release the manager and router changes on pypi, if you are planning on doing that soon, since its useful for most people. If not, then just merge it whenever and I will plan on working on the management commands later this week.

bronsen commented on docs/components/IDGeneration.md:
```diff
@@ -22,6 +22,7 @@ The two included in the package are:
 
```
Shouldn't it say
> The _three_ included in the package are


Also, I wonder if one could clarify what is enumerated here, since all these are mentioned above: strategies, additional arguments, or requirements. (<-- this is me being very pedantic)

4word replied:
You are right, it should say three. I will update.

What do you mean clarify what is enumerated? There are no additional requirements other than what is listed, which basically includes Postgres and defining a shard epoch time. All of the exact requirements are in the Using section for this feature, which is in the ShardingAModel.md file. If thats not what you mean, let me know!

JBKahn commented:
I'm going to try looking at this tonight. I'll probably take a stab at writing the management commands I'd like to include and then we can see about releasing this ASAP since I've been really lazy about this.

JBKahn commented:
@4word if I write some commands, and this is reviewed, are any pieces missing or is this good to go (both from the point of view of the pull request and for a new release) from your point o f view?

4word commented:
All the pieces are there, should be good to be merged in.

JBKahn commented on django_sharding_library/fields.py:
```diff
@@ -156,3 +166,64 @@ class ShardForeignKeyStorageField(ShardForeignKeyStorageFieldMixin, ForeignKey):
+        # If the database needs similar types for key fields however, the only
+        # thing we can do is making AutoField an IntegerField.
+        rel_field = self.target_field
+        if rel_field.get_internal_type() is "BigIntegerField":
```
Did you mean to put `is` and not `==` for the string comparison?

JBKahn commented on django_sharding_library/fields.py:
```diff
@@ -156,3 +166,64 @@ class ShardForeignKeyStorageField(ShardForeignKeyStorageFieldMixin, ForeignKey):
+        # If the database needs similar types for key fields however, the only
+        # thing we can do is making AutoField an IntegerField.
+        rel_field = self.target_field
+        if rel_field.get_internal_type() is "BigIntegerField":
```
ditto

JBKahn commented:
@4word those are my last two questions, then I should be good to merge this in and write the commands.

JBKahn commented:
I'm going to merge and make the change and let me know if I should revert it. That way I can continue working on it before you get back to me.
